Cleanup and crash fix around UnixSignal.WaitAny (bug #36214)

xmcclure opened the pull request:
Issue 1: Do not fsync on a pipe(). That is invalid, and apparently corrupts the stack about one out of every 100,000 tries on current OS X. (This is the one that was occurring on the builders, and resulted in SIGSEGV)

Issue 2: Currently there is a race condition in signal.c where a signal-waiting thread can close the pipe and then the signal handler can try to write to it. Introduce a locking mechanism to prevent this. (I coaxed this one into happening in standalone testing; it caused SIGPIPE)

In addition, some general cleanup:

signal.c: More comments, and use of atomic read and write functions is now more consistent.
nunit: Improve error reporting in case of a test returning a null results object in the absence of an exception (this might only occur in case of memory corruption, but I definitely saw it).

xmcclure commented on support/signal.c:
```diff
@@ -212,42 +303,52 @@ Mono_Unix_UnixSignal_install (int sig)
-		mph_int_set (&h->pipecnt, h->pipecnt, 0);
+		// If we reached here without have_restore_handler, this means that default_handler
+		// was set as the signal handler before the first UnixSignal object was installed.
+		g_assert (have_restore_handler);
```
Notice I did add an assert here.

kumpera commented on support/map.h:
```diff
@@ -2014,13 +2014,14 @@ Mono_Posix_ToUtimbuf (struct utimbuf *from, struct Mono_Posix_Utimbuf* to);
+	int   write_fd; // Read by signal handlers
+	int   pipecnt;  // Read by signal handlers
+	int   pipelock; // Used by pipelock_ functions in signal.c. Read and written by signal handlers
+	int   have_restore_handler;
```
restore_handler and have_restore_handler are simply renames. Let's remove this part.

kumpera commented on support/signal.c:
```diff
@@ -212,42 +303,52 @@ Mono_Unix_UnixSignal_install (int sig)
-				signals [i].handler != default_handler) {
-			have_handler = 1;
-			handler = signals [i].handler;
+		// Check if this slot has a "usable" (not installed by this file) restore handler
```
What's a restore handler? We're just looking at the previous handler and being nice unix citizens by reinstalling the old one when we're done.

kumpera commented:
There's a pair of variable renames that should be taken out first before this commit is ready.

xmcclure commented:
@jonp: What I know is that in my testing, default_handler was definitely running many times concurrently (I had it incrementing a static variable on entry and decrementing on exit, and it got up to, I think, 15 at the highest I saw it).

xmcclure commented:
Updated to back out the variable name change and a couple of unhelpful comments. Is this good to check in, or are there problems with some of the new atomic operations, @jonpryor?

jonpryor commented on support/map.h:
```diff
@@ -2014,13 +2014,14 @@ Mono_Posix_ToUtimbuf (struct utimbuf *from, struct Mono_Posix_Utimbuf* to);
 struct Mono_Unix_UnixSignal_SignalInfo {
-	int   signum;
-	int   count;
+	int   signum;   // Read by signal handlers
```
These comments are pointless; `map.h` is a generated file (see comment at top of file), and thus the comments will be lost the next time `create-native-map` is run.

If these comments are needed, they should be placed in the C# declaration.

jonpryor commented on support/map.h:
```diff
@@ -2014,13 +2014,14 @@ Mono_Posix_ToUtimbuf (struct utimbuf *from, struct Mono_Posix_Utimbuf* to);
 	int   read_fd;
-	int   write_fd;
+	int   write_fd; // Read by signal handlers
+	int   pipecnt;  // Read by signal handlers
```
...which brings us to ordering: the ordering (and naming) of fields here _must_ match that of `Mono.Unix.UnixSignal.SignalInfo`, in `UnixSignal.cs`.

You may also need to consider synchronization between the C# and C sides of this, e.g. C# accesses `SignalInfo.count`.

jonpryor commented on mcs/class/Mono.Posix/Test/Mono.Unix/UnixSignalTest.cs:
```diff
@@ -456,6 +456,7 @@ static void CloseSignals (UnixSignal[] signals)
 				s.Close ();
 		}
 
+		// Create thread that issues many signals from a set of harmless signals
```
Needless comment.

jonpryor commented on mcs/class/Mono.Posix/Test/Mono.Unix/UnixSignalTest.cs:
```diff
@@ -486,6 +487,7 @@ public void TestAddRemove ()
 			CloseSignals (usignals);
 		}
 
+		// Create thread that repeatedly registers then unregisters signal handlers
```
Needless comment.

jonpryor commented on mcs/class/Mono.Posix/Test/Mono.Unix/UnixSignalTest.cs:
```diff
@@ -523,6 +525,7 @@ public void TestWaitAny ()
 			CloseSignals (usignals);
 		}
 
+		// Create thread that blocks until at least one of the given signals is received
```
Needless comment.

jonpryor commented on support/signal.c:
```diff
@@ -175,30 +260,36 @@ default_handler (int signum)
+
 		fd = mph_int_get (&h->write_fd);
-		if (fd > 0) {
+		if (fd > 0) { // If any pipe exists to write to
```
Needless comment.

jonpryor commented on support/signal.c:
```diff
@@ -175,30 +260,36 @@ default_handler (int signum)
 			int j,pipecounter;
-			char c = signum;
-			pipecounter = mph_int_get (&h->pipecnt);
+			char c = signum;                         // (Actual value written to pipe is discarded)
```
Needless comment; what does this comment add?

We could instead do `char c = 0; /* value meaningless */`, but again, that doesn't necessarily add much.

What it *does* add is noise to the patch.

jonpryor commented on support/signal.c:
```diff
@@ -175,30 +260,36 @@ default_handler (int signum)
-	signal_info* h = NULL; 
-	int have_handler = 0;
+	signal_info* h = NULL;        // Info structure to save to
+	int have_handler = 0;         // Candidates for signal_info handler fields
```
Needless comments.

jonpryor commented on support/signal.c:
```diff
@@ -212,12 +303,15 @@ Mono_Unix_UnixSignal_install (int sig)
 			pthread_mutex_unlock (&signals_mutex);
 			errno = EADDRINUSE;
-			return NULL;
+			return NULL; // This is an rt signal with an existing handler. Bail out.
```
Needless comment.

jonpryor commented on support/signal.c:
```diff
@@ -290,11 +391,11 @@ Mono_Unix_UnixSignal_uninstall (void* info)
 			mph_sighandler_t p = signal (h->signum, h->handler);
 			if (p != SIG_ERR)
-				r = 0;
+				r = 0; // Success
```
Needless comment; context (e.g. the default value being -1), along with use in `UnixSignal`, show that this method is using Unix semantics -- -1 on error, 0 on success.

jonpryor commented on support/signal.c:
```diff
@@ -317,21 +419,22 @@ setup_pipes (signal_info** signals, int count, struct pollfd *fd_structs, int *c
 		h = signals [i];
 
-		if (mph_int_get (&h->pipecnt) == 0) {
+		if (h->pipecnt == 0) { // First pipe for this signal_info
```
This shouldn't be changed.

Upon further consideration, here's another two "design guidelines" for this code, that at least appear to partially be present in the existing code ;-)

1. If a member is shared between `default_handler()` and *anything else*, it should *probably* be atomic. This would support your previous change to use `mph_int_set()` to update `h->signum` (above), as `h->signum` is also used in `default_handler()`.
2. If a member is used atomically anywhere, it should be used atomically *everywhere*, so as to stop making my head hurt regarding whether the atomic read is actually necessary. (Extra use of atomics is *fine*, except for additional performance issues, which likely aren't an issue here. Not enough use of atomics can break everything. Consequently, the "stop making my head hurt" answer is "be consistent.")

jonpryor commented on support/signal.c:
```diff
@@ -317,21 +419,22 @@ setup_pipes (signal_info** signals, int count, struct pollfd *fd_structs, int *c
 		fd_structs[*currfd].fd = h->read_fd;
 		fd_structs[*currfd].events = POLLIN;
-		++(*currfd);
+		++(*currfd); // count is verified less than NUM_SIGNALS by caller
```
I don't think this comment is needed.

jonpryor commented on support/signal.c:
```diff
@@ -317,21 +419,22 @@ setup_pipes (signal_info** signals, int count, struct pollfd *fd_structs, int *c
 	return r;
 }
 
+// Cleanup a signal_info after waiting for signal
```
Unnecessary comment.

jonpryor commented on support/signal.c:
```diff
@@ -339,21 +442,26 @@ teardown_pipes (signal_info** signals, int count)
 		signal_info* h = signals [i];
 
-		if (mph_int_dec_test (&h->pipecnt)) {
+		if (mph_int_dec_test (&h->pipecnt)) { // Final pipe for this signal_info
```
Unnecessary comment.

jonpryor commented on support/signal.c:
```diff
@@ -166,6 +167,90 @@ keep_trying (int r)
+// This tiny ad-hoc read/write lock is needed because of the very specific
+// synchronization needed between default_handler and teardown_pipes:
+// - Many default_handlers can be running at once
+// - External forces already ensure only one teardown_pipes runs at once
```
Might be good to elaborate upon "external forces", e.g. that `teardown_pipes()` is only invoked when the `signals_mutex` mutex is held, so it *can't* be run concurrently.

jonpryor commented on support/signal.c:
```diff
@@ -166,6 +167,90 @@ keep_trying (int r)
+// synchronization needed between default_handler and teardown_pipes:
+// - Many default_handlers can be running at once
+// - External forces already ensure only one teardown_pipes runs at once
+// - If a teardown_pipes interrupts a default_handler, it must block
```
I may be very naive, but *how* can `teardown_pipes` interrupt `default_handler` execution? What does "interrupt" mean in this context?

I can imagine `teardown_pipes()` *executing concurrently* with `default_handler()` on two separate threads (and two separate CPUs), but that's not "interrupt[ing execution]"...

jonpryor commented on support/signal.c:
```diff
@@ -339,21 +442,26 @@ teardown_pipes (signal_info** signals, int count)
 {
 	int r, idx;
+	// Poll until one of this signal_info's pipes is ready to read.
+	// Once a second, stop to check if the VM is shutting down.
```
Needless comment

jonpryor commented on support/signal.c:
```diff
@@ -368,7 +476,7 @@ wait_for_any (signal_info** signals, int count, int *currfd, struct pollfd* fd_s
 	if (r == 0)
 		idx = timeout;
-	else if (r > 0) {
+	else if (r > 0) { // The pipe[s] are ready to read.
```
Needless comment.

jonpryor commented on support/signal.c:
```diff
@@ -379,7 +487,7 @@ wait_for_any (signal_info** signals, int count, int *currfd, struct pollfd* fd_s
 				} while (keep_trying (r) && !shutting_down ());
 				if (idx == -1)
-					idx = i;
+					idx = i; // This is the first index we read (unless we're shutting down)
```
Needless comment, and I'm not sure how accurate it is. What does "shutting down" have to do with the index?

xmcclure commented:
Rebased, addressed the map.h problem, and split comments/cleanup into a separate patch from the lock change. Have not attempted to address other issues yet.

xmcclure commented:
Adjusted "cleanup" commit to use atomics everywhere (so the rule to remember is simple) to and remove/amend some of the comments jon mentioned.
